samza-core/src/main/java/org/apache/samza/clustermanager/AbstractContainerAllocator.java

```diff
@@ -121,18 +146,101 @@ public void run() {
+   *
+   * During the run() method, the thread sleeps for allocatorSleepIntervalMs ms. It then invokes assignResourceRequests,
+   * and tries to allocate any unsatisfied request that is still in the request queue {@link ResourceRequestState})
+   * with allocated resources.
+   * When {@code hostAffinityEnabled} is disabled, all allocated resources are buffered in the list keyed by "ANY_HOST".
+   * When {@code hostAffinityEnabled} is enabled, all allocated resources are buffered in the list keyed by "preferredHost
+   *
+   * If the requested host is not available, the thread checks to see if the request has expired. If it has expired
+   * then two cases are handled seperately
+   *
+   * Case 1: host-affinity is disabled, cancels the current request and issues another ANY_HOST request
+   * Case 2: host-affinity is enabled, looks for allocated resouces on ANY_HOST and issues a container start if available,
+   *         otherwise issues an ANY_HOST request
+   *
+   * In either of the scenarious if a {@code StandbyContainerManager} is present, the allocator transfers the request
+   * to it for checking StandByConstraints
    */
-  protected abstract void assignResourceRequests();
+  protected void assignResourceRequests() {
+    while (hasReadyPendingRequest()) {
+      SamzaResourceRequest request = peekReadyPendingRequest().get();
+      String processorId = request.getProcessorId();
+      String preferredHost = hostAffinityEnabled ? request.getPreferredHost() : ResourceRequestState.ANY_HOST;
+      Instant requestCreationTime = request.getRequestTimestamp();
+
+      LOG.info("Handling assignment request for Processor ID: {} on host: {}.", processorId, preferredHost);
+      if (hasAllocatedResource(preferredHost)) {
+
+        // Found allocated container on preferredHost
+        LOG.info("Found an available container for Processor ID: {} on the host: {}", processorId, preferredHost);
+
+        // Needs to be only updated when host affinity is enabled
+        if (hostAffinityEnabled) {
+          state.matchedResourceRequests.incrementAndGet();
+        }
+
+        // Try to launch processor on this preferredHost if it all standby constraints are met
+        if (this.standbyContainerManager.isPresent()) {
+          standbyContainerManager.get().checkStandbyConstraintsAndRunStreamProcessor(request, preferredHost,
+              peekAllocatedResource(preferredHost), this, resourceRequestState);
+        } else {
+          runStreamProcessor(request, preferredHost);
+        }
+
+      } else {
+
+        LOG.info("Did not find any allocated containers for running Processor ID: {} on the host: {}.",
+            processorId, preferredHost);
+        boolean expired = isRequestExpired(request);
+
+        if (expired) {
+          updateExpiryMetrics(request);
+          if (hostAffinityEnabled) {
+            handleExpiredRequestWithHostAffinityEnabled(processorId, preferredHost, request);
+          }
```

Comment:
If hostAffinity is not enabled and the request has expired we need to do the following right?

`if (resourceAvailableOnAnyHost) {
      LOG.info("Request for Processor ID: {} on host: {} has expired. Running on ANY_HOST", processorId, preferredHost);
      runStreamProcessor(request, ResourceRequestState.ANY_HOST);
    } else {
      LOG.info("Request for Processor ID: {} on host: {} has expired. Requesting additional resources on ANY_HOST.",
          processorId, preferredHost);
      resourceRequestState.cancelResourceRequest(request);
      requestResource(processorId, ResourceRequestState.ANY_HOST);
    }`